crates/adapters/src/server/metrics.rs

```diff
@@ -0,0 +1,652 @@
+            }?;
+        }
+        Ok(())
+    }
+}
+
+/// Passed to [MetricsWriter::counter] and [MetricsWriter::gauge] callback.
+pub struct ValueWriter<'a, F>
+where
+    F: MetricsFormatter,
+{
+    formatter: &'a mut F,
+    name: &'a str,
+}
+
+impl<F> ValueWriter<'_, F>
+where
+    F: MetricsFormatter,
+{
+    /// Writes `value` as the value of this counter or gauge with the given
+    /// `labels`.  Each call to `write_value` for a given counter or gauge
+    /// should have different labels.
+    pub fn write_value(&mut self, labels: &LabelStack, value: impl Value) {
+        self.formatter
+            .write_value(self.name, "", labels, value.as_f64());
+    }
+}
+
+/// A type that can be written as the value of a metric.
+///
+/// All of these types are ultimately readable as a [f64].
+pub trait Value {
+    fn as_f64(&self) -> f64;
+}
+
+impl Value for f64 {
+    fn as_f64(&self) -> f64 {
+        *self
+    }
+}
+
+macro_rules! from_cast {
+    ($type_name:ty) => {
+        impl Value for $type_name {
+            fn as_f64(&self) -> f64 {
+                *self as f64
+            }
+        }
+    };
+}
+from_cast!(f32);
+from_cast!(u8);
+from_cast!(u16);
+from_cast!(u32);
+from_cast!(u64);
+from_cast!(u128);
+from_cast!(usize);
+from_cast!(i8);
+from_cast!(i16);
+from_cast!(i32);
+from_cast!(i64);
+from_cast!(i128);
+from_cast!(isize);
+
+macro_rules! from_atomic {
+    ($type_name:ty) => {
+        impl Value for &$type_name {
+            fn as_f64(&self) -> f64 {
+                self.load(Ordering::Relaxed) as f64
+            }
+        }
+    };
+}
+
+from_atomic!(AtomicU8);
+from_atomic!(AtomicU16);
+from_atomic!(AtomicU32);
+from_atomic!(AtomicU64);
+from_atomic!(AtomicI8);
+from_atomic!(AtomicI16);
+from_atomic!(AtomicI32);
+from_atomic!(AtomicI64);
+
+/// Passed to [MetricsWriter::histogram] callback.
+pub struct HistogramWriter<'a, F> {
+    formatter: &'a mut F,
+    name: &'a str,
+}
+
+impl<'a, F> HistogramWriter<'a, F>
+where
+    F: MetricsFormatter,
+{
+    /// Writes `histogram` as the value of this counter or gauge with the given
+    /// `labels`.  Each call to `write_histogram` for a given histogram should
+    /// have different labels.
+    pub fn write_histogram(&mut self, labels: &LabelStack, histogram: &impl Histogram) {
+        self.formatter.write_histogram(self.name, labels, histogram);
+    }
+}
+
+/// A bucket in a histogram.
+pub struct Bucket {
+    /// The upper limit of the bucket.
+    upper: f64,
+
+    /// The count in the bucket.
+    count: f64,
+}
+
+/// A histogram for passing to [HistogramWriter].
+pub trait Histogram {
+    /// Sum of all the buckets.
+    fn sum(&self) -> f64;
+    /// Total count in the histogram.
+    fn count(&self) -> f64;
+    /// All the buckets in the histogram.  The final bucket should have upper
+    /// limit [f64::INFINITY] and count equal to [count][Self::count].
+    fn buckets(&self) -> impl Iterator<Item = Bucket>;
```

Comment:
so these are not necessarily adjacent or equal-sized?

Reply:
No, the example at https://prometheus.io/docs/instrumenting/exposition_formats/ shows the buckets being roughly, but not exactly, exponential: 0.05, 0.1, 0.2, 0.5, 1, +Inf.